token-client/src/main/java/com/sap/cloud/security/xsuaa/client/AbstractOAuth2TokenService.java

```diff
@@ -1,21 +1,52 @@
 package com.sap.cloud.security.xsuaa.client;
 
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.Ticker;
 import com.sap.cloud.security.xsuaa.http.HttpHeaders;
 import com.sap.cloud.security.xsuaa.http.HttpHeadersFactory;
+import com.sap.cloud.security.xsuaa.tokenflows.CacheConfiguration;
+import com.sap.cloud.security.xsuaa.tokenflows.Cacheable;
 import com.sap.cloud.security.xsuaa.util.UriUtil;
 
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
 import java.net.URI;
 import java.util.Map;
+import java.util.Objects;
 
 import static com.sap.cloud.security.xsuaa.Assertions.assertNotNull;
 import static com.sap.cloud.security.xsuaa.client.OAuth2TokenServiceConstants.*;
 
 @java.lang.SuppressWarnings("squid:S1192")
-public abstract class AbstractOAuth2TokenService implements OAuth2TokenService {
+public abstract class AbstractOAuth2TokenService implements OAuth2TokenService, Cacheable {
+
+	private Cache<CacheKey, OAuth2TokenResponse> responseCache;
+
+	private final boolean sameThreadCache; // used for cache testing
+	private final Ticker ticker; // used for cache testing
 
 	public AbstractOAuth2TokenService() {
+		this(Ticker.systemTicker(), false);
+	}
+
+	/**
+	 * This constructor is used for testing purposes only.
+	 *
+	 * @param cacheTicker
+	 *            will be used in the cache to determine the time.
+	 * @param sameThreadCache
+	 *            set to true disables maintenance jobs of the cache. This makes the
+	 *            cache slower but more predictable for testing.
+	 */
+	AbstractOAuth2TokenService(Ticker cacheTicker, boolean sameThreadCache) {
+		ticker = cacheTicker;
+		this.sameThreadCache = sameThreadCache;
+	}
+
+	@Override
+	public void clearCache() {
+		getOrCreateResponseCache().invalidateAll();
```

Comment:
cache creation seems not to be required...?

if (Objects.nonNull(responseCache)) { responseCache.invalidateAll(); }

Reply:
I have reworked this part of the code, please have a look at it again.